### DISCUSSION TITLE
How to translate this piece of code to PascalABC.NET?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#2931] [Q&A] [answered]
```
using Eto.Forms;
using Eto.Drawing;

public class MyForm : Form
{
	public MyForm ()
	{
		Title = "My Cross-Platform App";
		ClientSize = new Size(200, 200);
		Content = new Label { Text = "Hello World!" };
	}
	
	[STAThread]
	static void Main()
	{
		new Application().Run(new MyForm());
	}
}
```

**Comment by SunSerega** [comment]
What part exactly do you not know how to translate? Start by showing your attempts in Pascal, that way you will learn more (through hands-on experience) and I will be able to make more relevant suggestions, including about things you just assumed cannot be done better.

P.S. ig you were that author of both https://github.com/pascalabcnet/pascalabcnet/issues/2928 and https://github.com/pascalabcnet/pascalabcnet/issues/2926? Why do you keep killing and recreating accounts?

**Reply by ghost** [reply]
Don't know how to translate the Object Initializer syntax to PascalABC.NET, especially when Inheritance is involved. Have read the CHM file but can't find the relevant information. I have tried my best and have failed. Do you want me to post my failed translations?

**Reply by ghost** [reply]
Github keeps suggesting me projects that I don't care about. Sometimes, three entries are the same project duplicated. It's annoying.

**Reply by ghost** [reply]
Sometimes, they populated my github.com page with activities from people that I don't care about. I contributed to various projects but it doesn't mean I care about them. I only think that I could help. Github suggested these projects to me. I have a quick look. Found a problem and reported back. Sometimes, if the issue is too trivial, I submitted a PR. Just that.

**Reply by SunSerega** [reply]
"Object Initializer" in OOP are called "constructor"'s. `constructor` is also a Pascal keyword, with its own F1 page. You can search for it in CHM, or you can write this word, place the cursor on it, and press F1 to open the right page.

> Sometimes, they populated my github.com page with activities from people that I don't care about

Are you saying it as your reason for multiple dead accounts? Well, why do you care about what you see on the main page anyway?
If you want to see updates from projects you are interested in - configure email notifications (in settings and on the pages of relevant projects). They would at least be thorough, unlike some compilations on the main page.

> Do you want me to post my failed translations?

Yes, preferably with comments about your thought process.

**Comment by ghost** [comment]
> "Object Initializer" in OOP are called "constructor"'s. `constructor` is also a Pascal keyword, with its own F1 page. You can search for it in CHM, or you can write this word, place the cursor on it, and press F1 to open the right page.

Object Initializer is a particular feature of C#. See MSDN: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer

**Reply by SunSerega** [reply]
Ah, you're right, I confused it.
Well, there are no inline property initializers in Pascal, for now, you'll need to assign `new Label` to an extra variable.

**Comment by ghost** [comment]
> Are you saying it as your reason for multiple dead accounts? Well, why do you care about what you see on the main page anyway? If you want to see updates from projects you are interested in - configure email notifications (in settings and on the pages of relevant projects). They would at least be thorough, unlike some compilations on the main page.

You don't know the number of emails I received. Hundreds of them. The number of projects I contributed is huge. Imagine you posted a new issue and received about 30 replies in the same day and you have many issues like that. I don't check the email box for this reason. I go straight to Github.

I'm a hobbyist. I do care about what I see on my github.com page. It is the source for me to find new projects to contribute to.

**Comment by ghost** [comment]
> Yes, preferably with comments about your thought process.

No comments, though. I don't see the needs for comments. I'm a hobbyist. Most of the code is throw away code anyway.

```
uses Eto.Forms, Eto.Drawing;

{$apptype windows}

type
  MyForm = class(Form)
  end;

begin
  var frm1 := new MyForm();
  frm1.Title := 'My Cross-Platform App';
  frm1.ClientSize := new Size(200, 200);
  //var lbn1 := new Label('Hello World');
  //MyForm.Content := lbn1;
  new Application().Run(new MyForm());
end.
```

```
uses Eto.Forms, Eto.Drawing;

{$apptype windows}

type
  MyForm = class(Form)
  public
    constructor Create();
  end;

constructor MyForm.Create();
begin
  Title := 'My Cross-Platform App';
  //ClientSize := new Size(200, 200);
  //var lbn1 := new Label('Hello World');
  //Content := lbn1;
end;

begin
  Application().Run(new MyForm());
end.
```

**Reply by ghost** [reply]
The code involved `Label` is annoying. Tried to create it first with `var lbn1 := new Label();` and then set the text with `lbn1.Text = 'Hello World';`. Also failed.

**Reply by SunSerega** [reply]
I didn't mean comments in code, I meant the text here which comments on the whole code block.

> `new Application().Run(new MyForm());`

You need to either add another `()` to apply `.` strictly after the operator `new`, or use a different syntax to call the constructor:
```
Application.Create.Run(...);
```

> var lbn1 := new Label('Hello World');

There is no constructor that takes a string. You need to assign the `'Hello World'` to the same property (`Text`) as in the C# program.

**Comment by samuraiGH** [comment, marked as answer]
```pas
{$apptype windows}

uses 
  Eto.Forms,
  Eto.Drawing;

type
  MyForm = class (Form)
    public
      constructor;
      begin
        Title:= 'My Cross-Platform App';
	ClientSize:= new Size(200, 200);
        
        var l:= new &Label;
        l.Text:= 'Hello World!';
	Content:= l;
      end;
  end;

begin
  (new Application).Run(new MyForm);
end.
```

**Reply by ghost** [reply]
Stuck at `ClientSize:= new Size(200, 200);`.

**Reply by SunSerega** [reply]
You mean there is some specific error? What does it say?

**Reply by samuraiGH** [reply]
`Form` contains a property `.Size`?
then

`ClientSize:= new Eto.Drawing.Size(200, 200);`

**Reply by ghost** [reply]
Please help. Compilation is fine. But can't run.

`Project4.pas(22) : Runtime exception: Could not detect platform. Are you missing a platform assembly?`

**Reply by ghost** [reply]
I think it's #2912. I'm using WPF based Eto.Forms on a Console project.

**Reply by ghost** [reply]
The reason for using a Console project is to get rid of WinForms as I'm using WPF based Eto.Forms.

**Reply by SunSerega** [reply]
Yes, using a separate (non-WinForms) program is the right call. But `{$apptype windows}` already makes it a non-console app. So that's not the problem.

Regardless, you also omitted the exception type, but the error message is still easy to look up. You are, in fact, missing a platform assembly:
https://github.com/picoe/Eto/issues/1033

**Reply by ghost** [reply]
I don't think so. `Eto.Platform.Windows` is for WinForms. I already have `Eto.Platform.Wpf`. I should have everything needed. But it seems what is missing and not automatically referenced is the WPF assemblies themselves. So I think I was right. It's #2912.

**Reply by ghost** [reply]
I have browsed the `nuget` directory in my project. I have added all of the assemblies I found here (of course I choose the correct versions). But the error is still the same. I opened `Eto.Wpf.dll` in ILSpy and I found it referenced various assemblies, some assemblies I don't have, e.g: the assemblies related to WebView as I'm running Windows 7. `Eto.WinForms.dll` also referenced these WebView assemblies.

**Reply by ghost** [reply]
I also found the IDE doesn't automatically add all of the needed assemblies from downloaded nuget packages to the reference list. For example, `Eto.Wpf.dll` is not automatically added to the reference list even though I installed the `Eto.Platform.Wpf` package. I have to add it manually. Now the error changed from `Project4.pas(22) : Runtime exception: Could not detect platform. Are you missing a platform assembly?` to `Project4.pas(0) : Error by reading assembly 'C:\PABCWork.NET\Project4\Eto.Wpf.dll'`.

**Reply by SunSerega** [reply]
No, that doesn't make sense. The compiler doesn't change anything in the referenced assemblies. And if you have really referenced (and not just placed the .dll in the same folder) - it's not even an indirect reference. #2912 is about more complex cases.

When you get a runtime error, the "Errors" tab gets selected. But the "Output" tab should actually have far more info, including the stack trace. We can use this info to see exactly how Eto is choosing the platform, so post it here too.
In the worst case, if it's really an unsolvable and Pascal-only issue - it would be possible to intercept platform auto-selection and specify it manually. But it most certainly isn't.

**Reply by SunSerega** [reply]
>  I found it referenced various assemblies, some assemblies I don't have, e.g: the assemblies related to WebView

It would only raise an exception if the code in the assembly you do have - tries to take something for the referenced but missing assembly.

**Reply by ghost** [reply]
I don't say the compiler changed the referenced assemblies. Btw, I know what I do. I'm sure that I correctly referenced the assemblies. I use the IDE frontend to do so.

**Reply by ghost** [reply]
The Output tab is blank. The Compiler Messages tab has more information. But it's not useful, though. The same error as reported in the Error List tab:

```
[pabcnetc.exe]Read Eto.Wpf.dll... 
1 errors
Error by reading assembly 'C:\PABCWork.NET\Project4\Eto.Wpf.dll'
```

**Reply by ghost** [reply]
I'm pretty sure it's #2912.

**Reply by SunSerega** [reply]
The order of messages is not what I expected. The message about the `Runtime exception` changing to `compile-time error` got posted while I was writing about checking "Output" (which is only active during run time) and needing to reference the assembly. And GitHub didn't update it for me automatically.

But yes, now that you are getting an assembly reading error in the case of indirect references - it can be related to #2912. But also it sounds junky in other, unrelated ways, too... Well, I see you already posted #2932.
I'm probably also going to try some stuff tomorrow. As it is now - this discussion turned into remote guesswork.

**Comment by ghost** [comment]
> > I found it referenced various assemblies, some assemblies I don't have, e.g: the assemblies related to WebView
> 
> It would only raise an exception if the code in the assembly you do have - tries to take something for the referenced but missing assembly.

I will try to reference all of the assemblies listed by ILSpy that I already have and report back. But, not today. It's late night here on my place.

**Reply by ghost** [reply]
Due to circular dependency, it's impossible to add all of the referenced assemblies. I only added the top level assemblies. The assemblies referenced by these assemblies (expanded by clicking the + button in ILSpy) are not added. Doesn't work. Same error.

**Reply by ghost** [reply]
> > I found it referenced various assemblies, some assemblies I don't have, e.g: the assemblies related to WebView
> 
> It would only raise an exception if the code in the assembly you do have - tries to take something for the referenced but missing assembly.

I don't think so. At least for the PascalABC.NET compiler, this is not the case. It seems it requires all of the referenced assemblies to be available. Otherwise, it will keep giving you that error. The only thing I'm missing now is is the WebView assemblies.

**Reply by ghost** [reply]
Yeah, I confirmed this to be the case. I installed the nuget package `Microsoft.Web.WebView2` and it's running now.

**Reply by ghost** [reply]
This is how I do it. Start with the Windows Forms Application project type instead of Console Application and use WinForms based Eto instead of Wpf based Eto. After I installed all of the needed nuget packages and referenced all of the needed assemblies, it's running.

**Reply by ghost** [reply]
I still have no way to make the Wpf based Eto to work, though. It seems because I can't reference all of the needed WPF assemblies. #2933.

**Reply by ghost** [reply]
Now, I'm pretty sure that it's #2912 all the time.

**Reply by ghost** [reply]
I have the Wpf based Eto working, too. The key is using Windows Forms Application project type instead of Console Application.
